CRM/Core/DAO/Base.php

```diff
@@ -0,0 +1,302 @@
+      if ($field['usage']['import']) {
+        $field['import'] = TRUE;
+      }
+      $field['where'] = ((string) $definition->name) . ".$fieldName";
+      if (!empty($fieldXml->headerPattern)) {
+        $field['headerPattern'] = (string) $fieldXml->headerPattern;
+      }
+      if (!empty($fieldXml->dataPattern)) {
+        $field['dataPattern'] = (string) $fieldXml->dataPattern;
+      }
+      if ($field['usage']['export'] || (!$field['usage']['export'] && $field['usage']['import'])) {
+        $field['export'] = $field['usage']['export'];
+      }
+      if (!empty($fieldXml->contactType)) {
+        $field['contactType'] = (string) $fieldXml->contactType;
+      }
+      if (!empty($fieldXml->rule)) {
+        $field['rule'] = (string) $fieldXml->rule;
+      }
+      $permission = CRM_Utils_Schema::getFieldPermission($fieldXml);
+      if ($permission) {
+        $field['permission'] = $permission;
+      }
+      if (isset($fieldXml->default)) {
+        // Default value in the xml may or may not be quoted
+        $default = trim((string) $fieldXml->default, '"\'');
+        $field['default'] = $default === 'NULL' ? NULL : $default;
+      }
+      $field['table_name'] = (string) $definition->name;
+      $field['entity'] = (string) ($definition->entity ?? $definition->class);
+      $field['bao'] = $baoName;
+      $field['localizable'] = (int) ($fieldXml->localizable ?? 0);
+      if (!empty($fieldXml->localize_context)) {
+        $field['localize_context'] = (string) $fieldXml->localize_context;
+      }
+      $field += $foreignKeys[$fieldName] ?? [];
+      if (!empty($fieldXml->component)) {
+        $field['component'] = (string) $fieldXml->component;
+      }
+      if (!empty($fieldXml->serialize)) {
+        $field['unique_title'] = constant('CRM_Core_DAO::SERIALIZE_' . $fieldXml->serialize);
+      }
+      if (!empty($fieldXml->uniqueTitle)) {
+        $field['unique_title'] = static::_ts((string) $fieldXml->uniqueTitle);
+      }
+      if (!empty($fieldXml->deprecated)) {
+        $field['deprecated'] = TRUE;
+      }
+      $html = CRM_Utils_Schema::getFieldHtml($fieldXml);
+      if ($html) {
+        $field['html'] = $html;
+      }
+      $pseudoconstant = CRM_Utils_Schema::getFieldPseudoconstant($fieldXml);
+      if ($pseudoconstant) {
+        $field['pseudoconstant'] = $pseudoconstant;
+      }
+      if ($fieldName === $primaryKey || !empty($fieldXml->readonly)) {
+        $field['readonly'] = TRUE;
+      }
+      $field['add'] = CRM_Utils_Schema::toString('add', $fieldXml);
+      $fieldKey = (string) ($fieldXml->uniqueName ?? $fieldName);
+      $fields[$fieldKey] = $field;
+    }
+    CRM_Core_DAO_AllCoreTables::invoke(static::class, 'fields_callback', $fields);
+    return $fields;
+  }
+
+  private static function getEntityDefinition(): SimpleXMLElement {
+    if (!isset(Civi::$statics[static::class][__FUNCTION__])) {
+      [, $dir, , $name] = explode('_', static::class);
```

Comment:
Related edge cases:

* `CRM_Foo_BAO_Bar extends CRM_Foo_DAO_Bar` ==> Would appear as two different classes
* `CRM_Event_Cart_DAO_*` and `CRM_Mailing_Event_DAO_*` ==> Placement of "DAO"/"BAO" can vary
* In a test-suite, the schema/class-structure remains constant, but `Civi::$statics` is reset several thousand times
    * (`ClassScanner` has a kind of similar consideration where it caches class-structure. It uses `static::$caches` instead of `Civi::$statics`. Spot-checking `api_v3_ContactTest`, this saves ~7% on execution-time -- which should translate to 2-3min on each of the bigger suites.)

I suppose it's not complicated to address any of those (*though maybe less pretty*).

(I think similar notes would apply in this branch or similar variations with different file-formats.)

Reply:
Noted. It looks like we won't be using this code anyway given that the discussion has evolved to favor a different file format & loading mechanism. But still good points.